### PR TITLE
Encode unpickle call directly on context

### PR DESCRIPTION
**Pull request by nicolasstucki**
This removes the internal methods `PickledQuoted.{unpickleExpr, unpickleType}`

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Part of #10222

File: compiler/src/dotty/tools/dotc/quoted/QuoteContextImpl.scala [hunk 1]
```diff
@@ -2628,13 +2628,13 @@ class QuoteContextImpl private (ctx: Context) extends QuoteContext, scala.intern
 
   end reflect
 
-  def unpickleExpr(pickledQuote: PickledQuote): scala.quoted.Expr[Any] =
+  def unpickleExpr[T](pickledQuote: PickledQuote): scala.quoted.Expr[T] =
     val tree = PickledQuotes.unpickleTerm(pickledQuote)(using reflect.rootContext)
-    new scala.internal.quoted.Expr(tree, hash)
+    new scala.internal.quoted.Expr(tree, hash).asInstanceOf[scala.quoted.Expr[T]]
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/quoted/QuoteContextImpl.scala:2633]: It seems to me that the original type signature is better. By returning the type `Expr[T]`, there are no guarantees.

**Review by liufengyun — APPROVED** [review]
Otherwise, LGTM

File: compiler/src/dotty/tools/dotc/quoted/QuoteContextImpl.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/quoted/QuoteContextImpl.scala:2633]: Indeed, there are no guarantees. Calls to this method should only be inserted be the compiler.
The previous version added the cast after each call to this method, effectively having the same guaranties.

I will add some extra documentation to make it clear that no one should be calling this method directly. Currently, it is hidden from the user as it is defined in `QuoteContextInteranal`.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/quoted/QuoteContextImpl.scala:2633]: In general, all methods in `QuoteContextInteranal` should only be called by code generated by the compiler or something in the library.

**Pull request merged**